pkg/controller/restic_repository_controller.go

```diff
@@ -0,0 +1,227 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/heptio/ark/pkg/apis/ark/v1"
+	arkv1client "github.com/heptio/ark/pkg/generated/clientset/versioned/typed/ark/v1"
+	informers "github.com/heptio/ark/pkg/generated/informers/externalversions/ark/v1"
+	listers "github.com/heptio/ark/pkg/generated/listers/ark/v1"
+	"github.com/heptio/ark/pkg/restic"
+)
+
+type resticRepositoryController struct {
+	*genericController
+
+	resticRepositoryClient arkv1client.ResticRepositoriesGetter
+	resticRepositoryLister listers.ResticRepositoryLister
+	bucket                 string
+	resticMgr              restic.RepositoryManager
+
+	clock clock.Clock
+}
+
+// NewResticRepositoryController creates a new restic repository controller.
+func NewResticRepositoryController(
+	logger logrus.FieldLogger,
+	resticRepositoryInformer informers.ResticRepositoryInformer,
+	resticRepositoryClient arkv1client.ResticRepositoriesGetter,
+	bucket string,
+	resticMgr restic.RepositoryManager,
+) Interface {
+	c := &resticRepositoryController{
+		genericController:      newGenericController("restic-repository", logger),
+		resticRepositoryClient: resticRepositoryClient,
+		resticRepositoryLister: resticRepositoryInformer.Lister(),
+		bucket:                 bucket,
+		resticMgr:              resticMgr,
+		clock:                  &clock.RealClock{},
+	}
+
+	c.syncHandler = c.processQueueItem
+	c.cacheSyncWaiters = append(c.cacheSyncWaiters, resticRepositoryInformer.Informer().HasSynced)
+
+	resticRepositoryInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.enqueue,
+			UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
+		},
+	)
+
+	c.resyncPeriod = 30 * time.Minute
+	c.resyncFunc = c.enqueueAllRepositories
+
+	return c
+}
+
+// enqueueAllRepositories lists all restic repositories from cache and enqueues all
+// of them so we can check each one for maintenance.
+func (c *resticRepositoryController) enqueueAllRepositories() {
+	c.logger.Debug("resticRepositoryController.enqueueAllRepositories")
+
+	repos, err := c.resticRepositoryLister.List(labels.Everything())
+	if err != nil {
+		c.logger.WithError(errors.WithStack(err)).Error("error listing restic repositories")
+		return
+	}
+
+	for _, repo := range repos {
+		c.enqueue(repo)
+	}
+}
+
+func (c *resticRepositoryController) processQueueItem(key string) error {
+	log := c.logger.WithField("key", key)
+	log.Debug("Running processItem")
+
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		log.WithError(errors.WithStack(err)).Error("error splitting queue key")
+		return nil
+	}
+
+	req, err := c.resticRepositoryLister.ResticRepositories(ns).Get(name)
+	if apierrors.IsNotFound(err) {
+		log.Debug("Unable to find ResticRepository")
+		return nil
+	}
+	if err != nil {
+		return errors.Wrap(err, "error getting ResticRepository")
+	}
+
+	switch req.Status.Phase {
```

Comment:
Maybe check if the phase is failed & return early. Else keep going with a single DeepCopy() and then do this switch?